### PR TITLE
[JIT] Propagate requires_grad to autodiff subgraphs

### PR DESCRIPTION
**Pull request by davidberard98**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #71666

When JIT autodiff is constructing a gradient computation graph, it will only add gradients for tensors that require_grad. Previously, require_grad information was **not** propagated to the subgraph that autodiff used; as a result, autodiff would calculate *all* gradients, even if requires_grad had never been set during profiling runs. In certain cases, this can lead to performance issues. For example, during training, the gradient of the input data is not needed, but is still computed.

This propagates requires_grad to the subgraph passed into autodiff, so that autodiff will not compute unnecessary gradients.

Test: `./bin/test_jit --gtest_filter="AutodiffRemoveUnusedGradientsTest.Linear"`

Differential Revision: [D33725304](https://our.internmc.facebook.com/intern/diff/D33725304)

### CODE REVIEW

**Comment by davidberard98** [comment]
@davidberard98 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33725304).

**Comment by davidberard98** [comment]
@davidberard98 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33725304).

**Comment by davidberard98** [comment]
@davidberard98 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33725304).

File: torch/csrc/jit/runtime/profiling_graph_executor_impl.cpp [hunk 1]
```diff
@@ -222,6 +222,9 @@ bool guardDifferentiableGraph(Node* dnode) {
         }
       }
 
+      bool requires_grad = dni->type()->expectRef<TensorType>().requires_grad();
```
Comment (eellison) [inline torch/csrc/jit/runtime/profiling_graph_executor_impl.cpp:225]: this probably doesn't matter much, but `requires_grad` will return a `True` if requires_grad is `c10::nullopt` on the tensor type. `requiresGrad` preserves the optionality. We should probably use that api, just for clarity, even though later down the road if we don't know if a Value requires grad (c10::nullopt case) we will just assume that it does as you are doing here.

Comment (eellison) [inline torch/csrc/jit/runtime/profiling_graph_executor_impl.cpp:225]: nit: maybe add comment that because we are inserting a `RequiresGrad` check, the input's grad property is guaranteed, but other properties are not ?

**Review by eellison — APPROVED** [review]
Nice!! Great tests!!

File: test/cpp/jit/test_autodiff.cpp [hunk 2]
```diff
@@ -281,5 +284,73 @@ TEST(AutodiffTest, DifferentiateWithRequiresGrad) {
...
+      break;
+    }
+  }
+  ASSERT_NE(nullptr, diff_graph_node);
+
+  auto backward_graph = diff_graph_node->g(attr::ReverseSubgraph);
+
+  // "input" does not need gradient, so it shouldn't be in the backwards graph
+  testing::FileCheck()
+      .check_not("%grad_input")
```
[the remark this answers is not in the document]
Comment (davidberard98) [reply test/cpp/jit/test_autodiff.cpp:346]: Do you know how stable we expect the usage of matmul is, compared to the names of the variables? (e.g. it could be replaced with mm)

**Comment by davidberard98** [comment]
@davidberard98 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33725304).
